bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerCreateOp.java

```diff
@@ -200,4 +209,220 @@ private void createComplete(int rc, LedgerHandle lh) {
+            this.builderWriteQuorumSize = other.builderWriteQuorumSize;
+            this.builderPassword = other.builderPassword;
+            this.builderDigestType = other.builderDigestType;
+            this.builderCustomMetadata = other.builderCustomMetadata;
+            this.bk = other.bk;
+        }
+
+        @Override
+        public CreateAdvBuilder withLedgerId(long ledgerId) {
+            builderLedgerId = ledgerId;
+            return this;
+        }
+
+        @Override
+        public CompletableFuture<WriteAdvHandle> execute() {
+            CompletableFuture<WriteAdvHandle> counter = new CompletableFuture<>();
+            create(new SyncCreateCallback(), counter);
+            return counter;
+        }
+
+        private boolean validateAdv(CreateCallback cb, Object ctx) {
+             if (builderLedgerId != null && builderLedgerId < 0) {
+                cb.createComplete(BKException.Code.IncorrectParameterException, null, ctx);
+                return false;
+            }
+
+            if (builderWriteQuorumSize > builderEnsembleSize) {
+                cb.createComplete(BKException.Code.IncorrectParameterException, null, ctx);
+                return false;
+            }
+
+            if (builderAckQuorumSize > builderWriteQuorumSize) {
+                cb.createComplete(BKException.Code.IncorrectParameterException, null, ctx);
+                return false;
+            }
+
+            if (builderAckQuorumSize <= 0) {
+                cb.createComplete(BKException.Code.IncorrectParameterException, null, ctx);
+                return false;
+            }
+
+            if (builderPassword == null) {
+                cb.createComplete(BKException.Code.IncorrectParameterException, null, ctx);
+                return false;
+            }
+
+            if (builderDigestType == null) {
+                cb.createComplete(BKException.Code.IncorrectParameterException, null, ctx);
+                return false;
+            }
+
+            if (builderCustomMetadata == null) {
+                cb.createComplete(BKException.Code.IncorrectParameterException, null, ctx);
+                return false;
+            }
+
+            return true;
+        }
+
+        private void create(CreateCallback cb, Object ctx) {
+            if (!validateAdv(cb, ctx)) {
+                return;
+            }
+            LedgerCreateOp op = new LedgerCreateOp(bk, builderEnsembleSize,
+                    builderWriteQuorumSize, builderAckQuorumSize, DigestType.fromApiDigestType(builderDigestType),
+                    builderPassword, cb, ctx, builderCustomMetadata);
+            bk.getCloseLock().readLock().lock();
+            try {
+                if (bk.isClosed()) {
+                    cb.createComplete(BKException.Code.ClientClosedException, null, ctx);
+                    return;
+                }
+                op.initiateAdv(builderLedgerId == null ? -1L : builderLedgerId);
```

Comment:
builderLedgerId cann't be null, right? why this check here?

Reply:
builderLedgerId = null means use LedgerIdGenerator so it is legal. It is the default value of field builderLedgerId